### PR TITLE
Implement Raft Prototype Engine

### PR DESCRIPTION
**Pull request by aludvik**
This PR creates an initial prototype implementation of a Raft-based Sawtooth consensus engine.

### CODE REVIEW

File: src/engine.rs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+// Returns whether the engine should continue
+fn handle_update(node: &mut SawtoothRaftNode, update: Update) -> bool {
+    match update {
+        Update::BlockNew(block) => node.on_block_new(block),
+        Update::BlockValid(block_id) => node.on_block_valid(block_id),
+        Update::BlockCommit(block_id) => node.on_block_commit(block_id),
+        Update::PeerMessage(message, _id) => node.on_peer_message(message),
+        Update::Shutdown => return false,
+
+        // TODO: Handle invalid, peer update
```
Comment (peterschwarz) [inline src/engine.rs]: Remove this: TODO

File: src/node.rs [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+        if let Some(committed_entries) = ready.committed_entries.take() {
+            let mut _last_apply_index = 0;
+            for entry in committed_entries {
+                // Mostly, you need to save the last apply index to resume applying
+                // after restart. Here we just ignore this because we use a Memory storage.
+                _last_apply_index = entry.get_index();
+
+                if entry.get_data().is_empty() {
+                    // Emtpy entry, when the peer becomes Leader it will send an empty entry.
```
Comment (peterschwarz) [inline src/node.rs]: s/Emtpy/Empty

File: src/node.rs [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+                    continue;
+                }
+
+                if entry.get_entry_type() == EntryType::EntryNormal {
+                    // TODO: This usage of Raft only proposes one block at a time. Performance
```
Comment (peterschwarz) [inline src/node.rs]: Remove TODO

File: src/node.rs [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+                        },
+                        _ => false,
+                    } {
+                        debug!("Leader({}) transitioning to Committing block {:?}", self.id, block_id);
+                        self.leader_state = Some(LeaderState::Committing(block_id.clone()));
+                    }
+                    self.service.commit_block(block_id.clone()).expect("Failed to commit block");
+                }
+
+                // TODO: handle EntryConfChange
```
Comment (peterschwarz) [inline src/node.rs]: Remove TODO.  If this is a remaining, unhandled case, a comment describing the case may be helpful

File: README.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+with `sawtooth.consensus.raft`.
+
+The only required setting is `sawtooth.consensus.raft.peers`. It must contain
+a mapping of each node's public key to each node's raft id.
+
+### Required Settings
+
+| key | value |
+| --- | --- |
+| peers | JSON - Map<PeerId, u64> |
```
[the remark this answers is not in the document]
Comment (peterschwarz) [reply README.md]: The json is submitted in the settings transaction, not computed by a transaction processor.  It is deterministic in this case in the same way that a timestamp submitted in a transaction body is deterministic (with respect to validation).

[the remark this answers is not in the document]
Comment (aludvik) [reply README.md]: @peterschwarz is right that once this payload is constructed, it isn't going to change. We could run into a problem if we try to iterate over the map produced and expect that iteration to be non-deterministic.

@dcmiddle do you have any thoughts on taking the last 64 bits of the peer id (which is the validator's public key) and casting that to a u64 to use as the Raft Id. This map was the simplest thing to do to get the prototype working, but if I am able to derive a unique u64 from each validator's public key, then this just needs to be a list of PeerIds (validator public keys) and the need to work with Raft Ids directly goes away.

[the remark this answers is not in the document]
Comment (dcmiddle) [reply README.md]: @aludvik Yes that's a good idea. The lower 64 bytes of a serialized secp256k1 public key (uncompressed) will do what you want. The top byte is just a bitcoin key format flag. If we were to switch curves to something else we could always hash into a 64 byte field without any real concern of collisions. An Edwards pubkey for example would only be 32 bytes so we could either pad out 0's on the rest of the bytes or hash it into 64 bytes.

**Pull request merged**